examples/llm_compression/torch/qat_with_nls_downstream/main.py

```diff
@@ -0,0 +1,773 @@
+        type=float,
+        default=1e-4,
+        help="Learning rate for fine-tuning. "
+        "For larger models (over 2 billion parameters), a learning rate of 5e-4 is recommended.",
+    )
+    parser.add_argument("--epochs", type=int, default=3, help="Number of epochs.")
+    parser.add_argument("--batch_size", type=int, default=16, help="Size of training batch.")
+    parser.add_argument(
+        "--microbatch_size",
+        type=int,
+        default=16,
+        help="Size of each training microbatch. Gradients will be accumulated until the batch size is reached.",
+    )
+    parser.add_argument("--eval_batch_size", type=int, default=64, help="Size of batch for evaluation.")
+
+    # Neural Low-rank Adapter Search (NLS) params
+    parser.add_argument(
+        "--lora_rank_space",
+        type=int,
+        nargs="+",
+        default=None,
+        help="Search space for LoRA adapter ranks. For example, if the (maximum) rank is 32, "
+        "this can be [32, 24, 16] to specify the ranks to be used during NLS.",
+    )
+    parser.add_argument(
+        "--custom_rank_config",
+        type=int,
+        nargs="+",
+        default=None,
+        help="Custom LoRA rank configuration (NLS) for evaluation.",
+    )
+    return parser
+
+
+def main(argv) -> float:
+    """
+    Fine-tuning and evaluating a language model with quantization-aware training and LoRA adapters,
+    including optional Neural Low-rank Adapter Search (NLS).
+    """
+    parser = get_argument_parser()
+    args = parser.parse_args(argv)
+    assert torch.cuda.is_available()
+    transformers.set_seed(42)
+    device = "cuda"
+    torch_dtype = torch.bfloat16
+    lora_rank = max(args.lora_rank_space)
+    disable_nls = len(args.lora_rank_space) == 1
+    compression_format = CompressionFormat.FQ_LORA if disable_nls else CompressionFormat.FQ_LORA_NLS
+    compression_config = dict(
+        mode=CompressWeightsMode.INT4_ASYM,
+        group_size=64,
+        compression_format=compression_format,
+        advanced_parameters=AdvancedCompressionParameters(lora_adapter_rank=lora_rank),
+    )
+
+    # Configure output and log files.
+    output_dir = Path(args.output_dir)
+    tensorboard_dir = output_dir / "tb" / datetime.now().strftime("%Y-%m-%d__%H-%M-%S")
+    last_dir = output_dir / "last"
+    ov_dir = output_dir / "ov"
+    result_file = output_dir / "result.json"
+
+    if not args.resume:
+        shutil.rmtree(output_dir, ignore_errors=True)
+    for path in [output_dir, tensorboard_dir, last_dir]:
+        path.mkdir(exist_ok=True, parents=True)
+    ckpt_file = last_dir / "nncf_checkpoint.pth"
+    print(f"To visualize the loss, open Tensorboard using the logs from: {tensorboard_dir}")
+    tb = SummaryWriter(tensorboard_dir, "QAT with absorbable LoRA")
+    overall_result = {}
+
+    # Load original model and tokenizer.
+    model = AutoModelForCausalLM.from_pretrained(args.pretrained, torch_dtype=torch_dtype, device_map=device)
```

Comment:
With default parameters, tuning can't fit to one RTX3090. What about using `device_map="auto"`? 
it would parallelize the model tuning on multiple GPU and default batch size will fit to 3 RTX3090.

Reply:
@ljaljushkin Done. Thanks!